### DISCUSSION TITLE
Confusing difference in session behaviour between 1.4 / 2.0

### DISCUSSION THREAD

**Discussion by Jackevansevo** [sqlalchemy/sqlalchemy#9565] [Usage Questions] [answered]
Hi all, I've recently been upgrading an application from SQLAlchemy 1.4 -> 2.0, and ran into some test failures. Specifically with code with the following pattern:

```python
user = session.get(User, user_id) # user is in session
person = Person()
person.user = user # person (by association) is also in session
session.flush() # should populate person.id
```

In `2.0` it appears that `person` is not added to the session by the assignment `person.user = user`, so the `session.flush()` ends up not populating `person.id`

Is anyone able to explain why this is now the case, I'd like to better understand what's going on here / specifically what's changed.

Apologies if I've completed missed something in the documentation / release notes.

---

I've managed to reduce my issue down to a minimal reproducible example. 

Running this script with `1.4` I get

```
user in session: True
person in session: True
person.id 1
```

But when running with `2.0` I get

```
user in session: True
person in session: False
person.id None
``` 

```python
from sqlalchemy import Column, ForeignKey, Integer, String, create_engine
from sqlalchemy.orm import Session, declarative_base, relationship

Base = declarative_base()


class Person(Base):
    __tablename__ = "person"

    id = Column(Integer, primary_key=True)
    user_id = Column(Integer, ForeignKey("user.id"), index=True)
    user = relationship("User", back_populates="persons")


class User(Base):
    __tablename__ = "user"

    id = Column(Integer, primary_key=True)
    name = Column(String)
    persons = relationship("Person", back_populates="user", cascade="all, delete")


engine = create_engine("sqlite://", echo=False)
Base.metadata.drop_all(engine)
Base.metadata.create_all(engine)

# Add user to DB
with Session(engine) as session:
    user = User(name="Jack")
    session.add(user)
    session.commit()

# Simulate what prod endpoint is doing
with Session(engine) as session:
    user = session.get(User, 1)
    print("user in session:", user in session)

    # By associated the person is now part of the session
    person = Person(user=user)
    print("person in session:", person in session)

    session.flush()

    # Will the person.id be populated???
    print("person.id", str(person.id))
```

**Comment by CaselIT** [comment]
Hi,

That's expected and it's documented in the 1.4 changelog, since IIRC it was warning / enabled when using `future=True`.
It's documented here https://docs.sqlalchemy.org/en/20/changelog/migration_14.html#cascade-backrefs-behavior-deprecated-for-removal-in-2-0

**Comment by zzzeek** [comment, marked as answer]
hey there -

so for the 1.4 to 2.0 transition, we have a [migration guide](https://docs.sqlalchemy.org/en/20/changelog/migration_20.html#x-2-x-migration-overview) that includes the `future=True` flag that will to the greatest degree possible have the engine/session behave as they would in 2.0, so that you can first get your program running under 1.4 in "2.0 compatibility" mode.   Additionally there's an envirionment variable SQLALCHEMY_WARN_20=1 that is part of this process which will emit warnings when known "changed in 2.0" things that we're able to catch occur in the program.

this particular behavioral change is one of the major changes that is covered by the above process and is documented in the 1.4 migration guide at https://docs.sqlalchemy.org/en/20/changelog/migration_14.html#cascade-backrefs-behavior-deprecated-for-removal-in-2-0 

So if we run your (well written, thanks!) program under 1.4.47 (I think as of maybe 1.4.45) you get a really prominent warning (it even uses colors):

```
/home/classic/dev/sqlalchemy/test3.py:39: RemovedIn20Warning: Deprecated API features detected! These feature(s) are not compatible with SQLAlchemy 2.0. To prevent incompatible upgrades prior to updating applications, ensure requirements files are pinned to "sqlalchemy<2.0". Set environment variable SQLALCHEMY_WARN_20=1 to show all deprecation warnings.  Set environment variable SQLALCHEMY_SILENCE_UBER_WARNING=1 to silence this message. (Background on SQLAlchemy 2.0 at: https://sqlalche.me/e/b8d9)
```

Hopefully you are seeing that warning on your end as well.   If we turn on the env variable SQLALCHEMY_WARN_20=1 we see a message and a link to the change:

```
/home/classic/dev/sqlalchemy/test3.py:39: RemovedIn20Warning: "Person" object is being merged into a Session along the backref cascade path for relationship "User.persons"; in SQLAlchemy 2.0, this reverse cascade will not take place.  Set cascade_backrefs to False in either the relationship() or backref() function for the 2.0 behavior; or to set globally for the whole Session, set the future=True flag (Background on this error at: https://sqlalche.me/e/14/s9r1) (Background on SQLAlchemy 2.0 at: https://sqlalche.me/e/b8d9)
  person = Person(user=user)
```

that link redirects to https://docs.sqlalchemy.org/en/14/errors.html#error-s9r1 which also describes the issue.


I guess what I'm getting at is, we went through a lot of effort to make this exact confusing thing complain **loudly** that it's happening, so I'm looking to see how these are still being missed.

**Reply by Jackevansevo** [reply]
Thanks so much for the response that's clarified everything for me now.

> I guess what I'm getting at is, we went through a lot of effort to make this exact confusing thing complain **loudly** that it's happening, so I'm looking to see how these are still being missed.

I'd been trying to reproduce the issue on `1.4.1` (same version as my app), so the warnings weren't appearing for me 🤦

**Reply by zzzeek** [reply]
OK taht's fair, I didnt add the more aggressive 1.4 warnings until much later
